go-sdk/stdlib/stdlib.go

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2021, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package stdlib
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"sync/atomic"
+)
+
+const (
+	// https://github.com/megaease/easemesh/blob/main/docs/sidecar-protocol.md#easemesh-traffic-hosting
+	agentPort = 9900
+)
+
+var (
+	agentAddr = fmt.Sprintf(":%d", agentPort)
+
+	// DefaultAgent is the default global agent.
+	DefaultAgent = NewAgent()
+)
+
+type (
+	// Agent is the agent entry.
+	Agent struct {
+		headers atomic.Value // type: []string
+	}
+
+	// AgentConfig is the config pushed to agent.
+	AgentConfig struct {
+		Headers string `json:"easeagent.progress.forwarded.headers"`
+	}
+
+	// AgentHandler is the HTTP handler wrapper.
+	AgentHandler struct {
+		handlerFunc http.HandlerFunc
+	}
+)
+
+// NewAgent returns the agent adapting part of EaseMesh sidecar protocol.
+func NewAgent() *Agent {
+	a := &Agent{}
+	a.headers.Store([]string{})
+	return a
+}
+
+// ServeDefault just runs global default agent in HTTP server,
+// please notice it prints logs if the server failed listening.
+// The caller must call it to activate default agent.
+func ServeDefault() {
+	go func() {
+		err := http.ListenAndServe(agentAddr, DefaultAgent)
+		if err != nil && err != http.ErrServerClosed {
+			log.Printf("easemesh agent listen %s failed: %v", agentAddr, err)
+		}
+	}()
+}
+
+// ServeHTTP serves function as agent such as health checking, config updating.
+func (a *Agent) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path == "/health" {
+		return
+	}
+
+	if r.URL.Path == "/config" {
+		a.handleConfig(w, r)
+	}
+}
+
+func (a *Agent) handleConfig(w http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("read config body failed: %v", err)
+		return
+	}
+
+	config := &AgentConfig{}
+	err = json.Unmarshal(body, config)
+	if err != nil {
+		log.Printf("unmarshal config body failed: %v", err)
+		return
+	}
+
+	headers := strings.Split(config.Headers, ",")
+	a.headers.Store(headers)
```

Comment:
Update only when the spec was changed?

Reply:
I have considered it, but it seems the operation to check if the config is changed won't be more efficient, moreover it requires to use more coarse-grained lock to compare-and-store. Considering that, I chose to simply store.

Reply:
It is fine with me